### PR TITLE
Create course offering editor

### PR DESCRIPTION
**Pull request by dmcavoy**
Creates the Course Offering Editor page. This is a page where editors can update the display name, category and if the course offering is featured in the category.

Screen Shot 2022-02-02 at 8 44 15 PM

## Links

[Assignment Dropdown Improvements](https://docs.google.com/document/d/11-dtCFuSmST_vP2MdkH0H0lb2e5AGFu1MxiWSRsnkns/edit#heading=h.3i8gd730jmak)

## Testing story

- Added unit tests
- Manually tested updating an existing course offering

### CODE REVIEW

**Comment by dmcavoy** [comment]
Make category and is_featured columns - make a migration to go in before this

File: dashboard/app/models/course_offering.rb [hunk 1]
```diff
@@ -72,6 +72,15 @@ def self.get_from_cache(key)
     end
   end
 
+  def summarize_for_edit
+    {
+      key: key,
+      is_featured: is_featured?,
```
Comment (bethanyaconnor) [inline dashboard/app/models/course_offering.rb:78]: I'm not 100% sure but I thought the pattern here is to use camelCase? So `isFeatured: is_featured?`. I might be wrong though -- this has always been an area I've been confused about.

Comment (dmcavoy) [reply dashboard/app/models/course_offering.rb:78]: +1 to the confusion. I have never figured out which way we are supposed to go

File: dashboard/app/controllers/course_offerings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  def i18n_params
+    params.permit(
+      :display_name
+    ).to_h
+  end
+
+  private
+
+  def course_offering_params
+    params.permit(:key, :display_name, :is_featured, :category).to_h
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/course_offerings_controller.rb]: Should `key` be included here? You use `params[:key]` above and I don't think we ever want to change the key, right?

File: dashboard/app/models/course_offering.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/course_offering.rb:78]: Yeah, definitely not something I'd block this PR on either way

File: dashboard/app/controllers/course_offerings_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    if Rails.application.config.levelbuilder_mode
+      @course_offering.write_serialization
+    end
+
+    render json: @course_offering.summarize_for_edit
+  rescue ActiveRecord::RecordNotFound, ActiveRecord::RecordInvalid => e
+    render(status: :not_acceptable, plain: e.message)
+  end
+
+  def i18n_params
```
Comment (bethanyaconnor) [inline dashboard/app/controllers/course_offerings_controller.rb]: What is this used for?

File: dashboard/test/controllers/course_offerings_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
+require 'test_helper'
+
+class CourseOfferingsControllerTest < ActionController::TestCase
```
Comment (bethanyaconnor) [inline dashboard/test/controllers/course_offerings_controller_test.rb:3]: I would add a test for the `update` function and making sure the correct fields are updated. I'm not sure how useful it is, especially with a simpler data model

File: apps/test/unit/lib/levelbuilder/CourseOfferingEditorTest.jsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+      server.respond();
+      wrapper.update();
+      expect(utils.navigateToHref).to.not.have.been.called;
+      expect(wrapper.find('.saveBar').find('FontAwesome').length).to.equal(0);
+      expect(
+        wrapper.find('.saveBar').contains('Error Saving: There was an error')
+      ).to.be.true;
+
+      server.restore();
```
Comment (bethanyaconnor) [inline apps/test/unit/lib/levelbuilder/CourseOfferingEditorTest.jsx]: nit: it looks like `server.restore()` is used in the `afterEach` function above (also applies to tests below)

File: apps/test/unit/lib/levelbuilder/CourseOfferingEditorTest.jsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+      ]);
+
+      const saveBar = wrapper.find('SaveBar');
+
+      const saveAndKeepEditingButton = saveBar.find('button').at(1);
+      expect(saveAndKeepEditingButton.contains('Save and Keep Editing')).to.be
+        .true;
+      saveAndKeepEditingButton.simulate('click');
+
+      // check the the spinner is showing
```
Comment (bethanyaconnor) [inline apps/test/unit/lib/levelbuilder/CourseOfferingEditorTest.jsx:61]: nit: I'm not sure that `CourseOfferingEditor` cares if the spinner is showing. The logic belongs to `SaveBar`

Longer ramble: I find it difficult to separate the concerns of `SaveBar` with the concerns of the editor (I had this problem with `ProgrammingExpressionEditor` as well). The test needs access to `SaveBar` to initiate saving but after that, I'm not sure what needs to be checked. Maybe the spinner being there should be checked, maybe it doesn't matter. 🤷‍♀️

Comment (dmcavoy) [reply apps/test/unit/lib/levelbuilder/CourseOfferingEditorTest.jsx:61]: Ya I also struggled. I have found figuring out how to test with the SaveBar complicated

File: dashboard/app/models/course_offering.rb (hunk 1 above)
Comment (dmcavoy) [reply dashboard/app/models/course_offering.rb:78]: It does allow for the cool thing where I can just saving the courseOffering in useState, update it, and then send it right back but I'm guessing I could quickly convert the keys somewhere in that process too

**Review by bethanyaconnor — APPROVED** [review]
Updates look good!

File: dashboard/app/models/course_offering.rb (hunk 1 above)
Comment (bethanyaconnor) [reply dashboard/app/models/course_offering.rb:78]: `transform_keys(&:underscore)` is useful for this, as well as the reverse `transform_keys(&:camelize)`. But it's still an extra step that doesn't always seem necessary (to me).

Comment (dmcavoy) [reply dashboard/app/models/course_offering.rb:78]: I'll leave it for now and can update if we decide thats how we want it going forward

**Pull request merged**